axelrod/strategies/finite_state_machines.py

```diff
@@ -1,7 +1,238 @@
 from axelrod.action import Action
 from axelrod.player import Player
+from typing import DefaultDict, Iterator, Dict, Tuple, Set, List
+from collections import defaultdict, namedtuple
 
 C, D = Action.C, Action.D
+ALL_ACTIONS = [C, D]
+
+
+"""
+Memit = unit of memory.
+
+This represents the amount of memory that we gain with each new piece of
+history.  It includes a state, our_response that we make on our way into that
+state (in_act), and the opponent's action that makes us move out of that state
+(out_act).
+
+For example for this FSM:
+(0, C, 0, C),
+(0, D, 1, C),
+(1, C, 0, D),
+(1, D, 0, D)
+
+Has the memits:
+(C, 0, C),
+(C, 0, D),
+(D, 0, C),
+(D, 0, D),
+(C, 1, C),
+(C, 1, D)
+"""
+Memit = namedtuple("Memit", ["in_act", "state", "out_act"])
+
+def memits_match(x: Memit, y: Memit):
+    """In action and out actions are the same."""
+    return x.in_act == y.in_act and x.out_act == y.out_act
+
+def memit_sort(x: Memit, y: Memit):
+    """Returns a tuple of x in y, sorted so that (x, y) are viewed as the
+    same as (y, x).
+    """
+    if repr(x) <= repr(y):
+        return (x, y)
+    else:
+        return (y, x)
+
+
+Transition = namedtuple("Transition", ["state", "last_opponent_action",
+                                       "next_state", "next_action"])
+TransitionDict = Dict[Tuple[int, Action], Tuple[int, Action]]
+
+def transition_iterator(transitions: TransitionDict) -> Iterator[Transition]:
+    """Changes the transition dictionary into a iterator on namedtuples, because
```

Comment:
`"""Changes the transition dictionary into a iterator on namedtuples."""`

Reply:
Done.